server/sonar-server/src/main/java/org/sonar/server/computation/step/NewCoverageAggregationStep.java

```diff
@@ -0,0 +1,250 @@
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.server.computation.step;
+
+import com.google.common.base.Function;
+import com.google.common.base.Optional;
+import com.google.common.collect.ImmutableList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import org.sonar.api.measures.CoreMetrics;
+import org.sonar.server.computation.component.Component;
+import org.sonar.server.computation.component.PathAwareVisitor;
+import org.sonar.server.computation.component.TreeRootHolder;
+import org.sonar.server.computation.measure.Measure;
+import org.sonar.server.computation.measure.MeasureRepository;
+import org.sonar.server.computation.measure.MeasureVariations;
+import org.sonar.server.computation.metric.Metric;
+import org.sonar.server.computation.metric.MetricRepository;
+import org.sonar.server.computation.period.Period;
+import org.sonar.server.computation.period.PeriodsHolder;
+
+import static com.google.common.collect.FluentIterable.from;
+import static org.sonar.server.computation.component.Component.Type.FILE;
+import static org.sonar.server.computation.component.ComponentVisitor.Order.POST_ORDER;
+import static org.sonar.server.computation.measure.Measure.newMeasureBuilder;
+
+public class NewCoverageAggregationStep implements ComputationStep {
+  private static final List<String> AGGREGATED_METRIC_KEYS = ImmutableList.of(
+    CoreMetrics.NEW_LINES_TO_COVER_KEY,
+    CoreMetrics.NEW_UNCOVERED_LINES_KEY,
+    CoreMetrics.NEW_CONDITIONS_TO_COVER_KEY,
+    CoreMetrics.NEW_UNCOVERED_CONDITIONS_KEY,
+    CoreMetrics.NEW_IT_LINES_TO_COVER_KEY,
+    CoreMetrics.NEW_IT_UNCOVERED_LINES_KEY,
+    CoreMetrics.NEW_IT_CONDITIONS_TO_COVER_KEY,
+    CoreMetrics.NEW_IT_UNCOVERED_CONDITIONS_KEY,
+    CoreMetrics.NEW_OVERALL_LINES_TO_COVER_KEY,
+    CoreMetrics.NEW_OVERALL_UNCOVERED_LINES_KEY,
+    CoreMetrics.NEW_OVERALL_CONDITIONS_TO_COVER_KEY,
+    CoreMetrics.NEW_OVERALL_UNCOVERED_CONDITIONS_KEY
+    );
+  private static final VariationBuildersFactory VARIATION_BUILDERS_FACTORY = new VariationBuildersFactory();
+
+  private final TreeRootHolder treeRootHolder;
+  private final PeriodsHolder periodsHolder;
+  private final MeasureRepository measureRepository;
+  private final List<Metric> metrics;
+
+  public NewCoverageAggregationStep(TreeRootHolder treeRootHolder, PeriodsHolder periodsHolder,
+    final MetricRepository metricRepository, MeasureRepository measureRepository) {
+    this.treeRootHolder = treeRootHolder;
+    this.periodsHolder = periodsHolder;
+    this.measureRepository = measureRepository;
+    this.metrics = from(AGGREGATED_METRIC_KEYS)
+      .transform(new Function<String, Metric>() {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Make this a named "static" inner class. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2694)